### ISSUE TITLE
Referencing ${opt:stage} from resource section breaks when using default stage

### ISSUE THREAD

**Issue by dpwrussell** [serverless/serverless#4942]
# This is a (Bug Report)

## Description

Usually when referencing `${opt:stage}` and calling `serverless deploy` without an explicit stage, `dev` is used by default. Using the `${opt:stage}` variable in some places works, but in `resources` does not seem to:

E.g. This works in a step function definition (and the resultant string has `dev` for the stage):
```
Resource: arn:aws:lambda:#{AWS::Region}:#{AWS::AccountId}:function:${self:service}-${opt:stage}-checkStatusSyncJob
```

but this does not:
```
resources:
  Outputs:
    writeEvent:
      Value:
        Ref: WriteEventLambdaFunction
      Export:
        Name: ${self:service}-${opt:stage}-LambdaWriteEvent
```

However if I manually specify `--stage dev` when I run deploy, then the variable works as expected in `resources`.

This seems inconsistent.

## Additional Data

* ***Serverless Framework Version you're using***: 1.26.1
* ***Operating System***: OSX 10.13.4
* ***Stack Trace***: N/A
* ***Provider Error messages***:
```
 Serverless Warning --------------------------------------

  A valid option to satisfy the declaration 'opt:stage' could not be found.


  Serverless Error ---------------------------------------

  Trying to populate non string value into a string for variable ${opt:stage}. Please make sure the value of the property is a string.
```

**Comment by HyperBrain** [comment]
Hi @dpwrussell . `opt:stage` intentionally only references the option set on the CLI. To get the effective stage, you should define it as custom variable as follows:
```
custom:
  stage: ${opt:stage, self:provider.stage}
  ...
```
This will use the option if given and the implicit stage defined in the provider otherwise. In your resources section (or anywhere in the serverless.yml) you should then reference the stage with `${self:custom.stage}`.

**Comment by dpwrussell** [comment]
If that is the case, I don't understand why `${opt:stage} _does_ resolve to `dev` in other sections when I did not specify an explicit stage on the command line. Is this plugin dependent?

**Comment by erikerikson** [comment]
@dpwrussell Can you provide configuration that reproduces this?

**Comment by dpwrussell** [comment]
Sure.

Perhaps this is actually a bug/feature of the step function plugin?
I've indicated the `${opt:stage}`s that worked and those which did not.

```YAML
service: test

provider:
  name: aws
  runtime: python3.6

functions:
  submitJob:
    handler: handler.submit_job
    environment:
      # This one does not work (as you expected)
      STAGE: ${opt:stage}

  checkStatusJob:
    handler: handler.check_status_job
    environment:
      # This one does not work (as you expected)
      STAGE: ${opt:stage}

stepFunctions:
  stateMachines:
    MySM:
      definition:
        StartAt: SubmitJob
        States:
          SubmitJob:
            Type: Task
            # This one works just fine!
            Resource: arn:aws:lambda:#{AWS::Region}:#{AWS::AccountId}:function:${self:service}-${opt:stage}-submitJob
            ResultPath: $.jobId
            Next: Wait10
          Wait10:
            Type: Wait
            Seconds: 10
            Next: GetJobStatus
          GetJobStatus:
            Type: Task
            # This one works just fine!
            Resource: arn:aws:lambda:#{AWS::Region}:#{AWS::AccountId}:function:${self:service}-${opt:stage}-checkStatusJob
            InputPath: $.jobId
            ResultPath: $.status
            Next: IsJobComplete
          IsJobComplete:
            Type: Choice
            Choices:
              - Variable: $.status
                StringEquals: FAILED
                Next: JobFailed
              - Variable: $.status
                StringEquals: SUCCEEDED
                Next: GetFinalJobStatus
            Default: Wait10
          JobFailed:
            Type: Fail
            Cause: Job Failed
            Error: DescribeJob returned FAILED
          GetFinalJobStatus:
            Type: Task
            # This one works just fine!
            Resource: arn:aws:lambda:#{AWS::Region}:#{AWS::AccountId}:function:${self:service}-${opt:stage}-checkStatusJob
            InputPath: $.jobId
            End: true

resources:
  Resources:
    MySMARN:
      Type: AWS::SSM::Parameter
      Properties:
        # This one does not work (as you expected)
        Name: /${self:service}/${opt:stage}/batch/MySMARN
        Type: String
        Value:
          Ref: MySMStepFunctionsStateMachine

plugins:
  - serverless-python-requirements
  - serverless-pseudo-parameters
  - serverless-step-functions
custom:
  pythonRequirements:
    dockerizePip: non-linux
```

**Comment by dpwrussell** [comment]
Ok, fair enough! Thanks.

**Comment by erikerikson** [comment]
@horike37 I'm curious: do you expect this option value backfill behavior?

**Comment by horike37** [comment]
What @HyperBrain expects is right.
The initialization process of Serverless Step Functions is different from the framework core.
It is accidentally, not intentionally.

Any definitions in serverless.yml by external plugins should be set inside `custom` section to initialize it by the framework core but I actually set it on the top level. Therefore, I created parsing process for `stepfunctions` section inside the plugin.

I want to fix it by moving the section but it is a breaking...
